### PR TITLE
feat: Added Periodicity to Consolidate Financial Statements and Cost Center Financial Statements and Added Variance feature to Consolidated Financial Statements

### PR DESCRIPTION
**Pull request by imdadhussain**
Task ID: https://bloomstack.com/desk#Form/Task/TASK-2021-00395

Screenshot/ GIFS

Periodicity GIF
period

Compare with Company Variance (A-B)GIF
varaince

Profit and losses GIF
profit and losses

Cost Center Financial Statement GIF
cost center

workflow:
1. Added Period List for balance_sheet_data. which call get_data and added the period key and company key 
2. Added company + Period Key for get_columns
3. company + period key and data generated by company + Period Key will append into data like above
4. Added and changes to get_profit_and_losses_data which call get_data and added period key and company in data as shown in GIF.
5. Made same changes for Cash Flow.
6. Made changes in Cost Center Financial Statement to make work as Consolidated Financial Statement because same method are in these and also attached Gif.

### CODE REVIEW

**Comment by neilLasrado** [comment]
@imdadhussain what's going on with this? The PR is in draft since 12 Days!

**Comment by imdadhussain** [comment]
> @imdadhussain what's going on with this? The PR is in draft since 12 Days!



> @imdadhussain what's going on with this? The PR is in draft since 12 Days!

@neilLasrado just created PR for reference for @vishdha

File: erpnext/accounts/report/consolidated_financial_statement/consolidated_financial_statement.py [hunk 1]
```diff
@@ -62,11 +68,16 @@ def get_balance_sheet_data(fiscal_year, companies, columns, filters, cost_center
 			"warn_if_negative": True,
 			"currency": company_currency
 		}
+		# for period in period_list:
+		# 	unclosed[period.key] = opening_balance
+		# 	if provisional_profit_loss:
+		# 		provisional_profit_loss[period.key] = provisional_profit_loss[period.key] - opening_balance
+		# unclosed["total"]=opening_balance
+
```
Comment (vishdha) [inline erpnext/accounts/report/consolidated_financial_statement/consolidated_financial_statement.py]: @imdadhussain  Remove unwanted comments

File: erpnext/accounts/report/consolidated_financial_statement/consolidated_financial_statement.py [hunk 2]
```diff
@@ -312,19 +364,44 @@ def prepare_data(accounts, fiscal_year, balance_must_be, companies, company_curr
...
+						row[f'{company}({period.key})'] = flt(d.get(f'{company}({period.key})', 0.0), 3)
+
+						if abs(row[f'{company}({period.key})']) >= 0.005:
+							# ignore zero values
+							has_value = True
+							total += flt(row[f'{company}({period.key})'])
+
+					row["has_value"] = has_value
+					row["total"] = total
+					row[f"{company}(total)"] = total
```
Comment (vishdha) [inline erpnext/accounts/report/consolidated_financial_statement/consolidated_financial_statement.py]: This code and lines 386 to 400 are kind of redundant you may optimize code here

File: erpnext/accounts/report/balance_sheet/balance_sheet.py [hunk 3]
```diff
@@ -137,25 +137,51 @@ def get_provisional_profit_loss(asset, liability, equity, period_list, company,
...
+					if provisional_profit_loss[key]:
+						has_value = True
+
+					total += flt(provisional_profit_loss[key])
+					#provisional_profit_loss["total"] = total
+					provisional_profit_loss[f"{company}(total)"] = total
+
+					total_row_total += flt(total_row[key])
+					#total_row["total"] = total_row_total
+					total_row[f"{company}(total)"] = total_row_total
```
Comment (vishdha) [inline erpnext/accounts/report/balance_sheet/balance_sheet.py]: This code and lines 167 to 184 are kind of redundant we can optimize this

File: erpnext/accounts/report/consolidated_financial_statement/consolidated_financial_statement.py [hunk 4]
```diff
@@ -182,17 +194,54 @@ def get_columns(companies):
...
+					})
+				if periodicity!="Yearly":
+					if not accumulated_in_group_company:
+						columns.append({
+							#"fieldname": "total",
+							"fieldname": f"{company}(total)",
+							"label": f'{company} Total',
+							"fieldtype": "Currency",
+							"width": 150
+						})
```
Comment (vishdha) [inline erpnext/accounts/report/consolidated_financial_statement/consolidated_financial_statement.py]: This also optimizes Lines 218 to 233 same

**Review by vishdha — CHANGES_REQUESTED** [review]

File: erpnext/accounts/report/consolidated_financial_statement/consolidated_financial_statement.py [hunk 5]
```diff
@@ -399,23 +448,38 @@ def get_additional_conditions(from_date, ignore_closing_entries, filters):
...
-				row[company] = 0.0
-
-			total_row.setdefault("total", 0.0)
-			total_row["total"] += flt(row["total"])
-			row["total"] = ""
+		#if not row.get("parent_account"):
+		for company in companies:
+			#total_row.setdefault(company, 0.0)
+			#total_row[company] += row.get(company, 0.0)
+			#row[company] = 0.0
```
Comment (vishdha) [inline erpnext/accounts/report/consolidated_financial_statement/consolidated_financial_statement.py]:
```suggestion
		
```

**Review by vishdha — CHANGES_REQUESTED** [review]

**Comment by neilLasrado** [comment]
@vishdha please review this

**Review by vishdha — APPROVED** [review]

**Review by sahil28297 — APPROVED** [review]
Approving and merging this as it has been tested locally by @vishdha

**Pull request merged**